lua_modules/gossip/gossip.lua

```diff
@@ -0,0 +1,275 @@
+    state.tickNodeState(destination);
+  end
+end
+-- luacheck: pop
+
+network.pickRandomNode = function()
+  if #gossip.config.seedList > 0 then
+    local randomListPick = node.random(1, #gossip.config.seedList);
+    utils.debug('Randomly picked: ' .. gossip.config.seedList[randomListPick]);
+    return gossip.config.seedList[randomListPick];
+  end
+  utils.debug(
+      'Seedlist is empty. Please provide one or wait for node to be contacted.');
+  return nil;
+end
+
+network.sendData = function(ip, data, sendType)
+  local outboundSocket = net.createUDPSocket();
+  data.type = sendType;
+  local dataToSend = sjson.encode(data);
+  data.type = nil;
+  outboundSocket:send(gossip.config.comPort, ip, dataToSend);
+  outboundSocket:close();
+end
+
+network.receiveSyn = function(ip, synData)
+  utils.debug('Received SYN from ' .. ip);
+  local update = utils.getMinus(synData, gossip.networkState);
+  local diff = utils.getMinus(gossip.networkState, synData);
+  network.updateNetworkState(update);
+  network.sendAck(ip, diff);
+end
+
+network.receiveAck = function(ip, ackData)
+  utils.debug('Received ACK from ' .. ip);
+  local update = utils.getMinus(ackData, gossip.networkState);
+  network.updateNetworkState(update);
+end
+
+network.sendAck = function(ip, diff)
+  local diffIps = '';
+  for k in pairs(diff) do diffIps = diffIps .. ' ' .. k; end
+  utils.debug('Sending ACK to ' .. ip .. ' with ' .. diffIps .. ' updates.');
+  network.sendData(ip, diff, constants.updateType.ACK);
+end
+
+-- luacheck: push no unused
+network.receiveData = function(socket, data, port, ip)
+  if gossip.networkState[ip] then
+    gossip.networkState[ip].state = constants.nodeState.UP;
+  end
+  local messageDecoded, updateData = pcall(sjson.decode, data);
+  if not messageDecoded then
+    utils.debug('Invalid JSON received from ' .. ip);
+    return;
+  end
+  local updateType = updateData.type;
+  updateData.type = nil;
+  if updateType == constants.updateType.SYN then
+    network.receiveSyn(ip, updateData);
+  elseif updateType == constants.updateType.ACK then
+    network.receiveAck(ip, updateData);
+  else
+    utils.debug('Invalid data comming from ip ' .. ip ..
+                    '. No valid type specified.');
+  end
+end
+-- luacheck: pop
+
+-- Constants
+
+constants.nodeState = {TICK = 1, UP = 0, SUSPECT = 2, DOWN = 3, REMOVE = 4};
+
+constants.defaultConfig = {
+  seedList = {},
+  roundInterval = 15000,
+  comPort = 5000,
+  debug = false
+};
+
+constants.comparisonFields = {'revision', 'heartbeat', 'state'};
+
+constants.updateType = {ACK = 'ACK', SYN = 'SYN'}
+
+constants.revFileName = 'gossip/rev.dat';
+
+-- Return
+
+gossip = {
+  started = false,
+  config = constants.defaultConfig,
+  setConfig = utils.setConfig,
+  start = state.start,
+  setRevFileValue = state.setRevFileValue,
+  networkState = {},
+  getNetworkState = utils.getNetworkState,
+  pushGossip = network.pushGossip
+};
+
+-- unit tests
+
```

Comment:
Having to change the sources does not work well together with automated testing.
Using something like this together with a change in the test might help
``` Lua
-- load with loadfile("gossip.lua")("test") to run gossip_tests
if (... == "test") then
  gossip._test = {
   _gossip = gossip,
   _constants = constants,
   _utils = utils,
   _network = network,
   _state = state
  }
end
```